src/cmap/connection.ts

```diff
@@ -757,3 +769,534 @@ function write(
+        completeCleanup();
+      });
+    } else {
+      completeCleanup();
+    }
+  }
+
+  command(
+    ns: MongoDBNamespace,
+    command: Document,
+    options: CommandOptions | undefined,
+    callback: Callback
+  ): void {
+    let cmd = { ...command };
+
+    const readPreference = getReadPreference(options);
+    const session = options?.session;
+
+    let clusterTime = this.clusterTime;
+
+    if (this.serverApi) {
+      const { version, strict, deprecationErrors } = this.serverApi;
+      cmd.apiVersion = version;
+      if (strict != null) cmd.apiStrict = strict;
+      if (deprecationErrors != null) cmd.apiDeprecationErrors = deprecationErrors;
+    }
+
+    if (this.hasSessionSupport && session) {
+      if (
+        session.clusterTime &&
+        clusterTime &&
+        session.clusterTime.clusterTime.greaterThan(clusterTime.clusterTime)
+      ) {
+        clusterTime = session.clusterTime;
+      }
+
+      const err = applySession(session, cmd, options);
+      if (err) {
+        return callback(err);
+      }
+    } else if (session?.explicit) {
+      return callback(new MongoCompatibilityError('Current topology does not support sessions'));
+    }
+
+    // if we have a known cluster time, gossip it
+    if (clusterTime) {
+      cmd.$clusterTime = clusterTime;
+    }
+
+    if (
+      // @ts-expect-error ModernConnections cannot be passed as connections
+      isSharded(this) &&
+      !this.supportsOpMsg &&
+      readPreference &&
+      readPreference.mode !== 'primary'
+    ) {
+      cmd = {
+        $query: cmd,
+        $readPreference: readPreference.toJSON()
+      };
+    }
+
+    const commandOptions: Document = Object.assign(
+      {
+        numberToSkip: 0,
+        numberToReturn: -1,
+        checkKeys: false,
+        // This value is not overridable
+        secondaryOk: readPreference.secondaryOk()
+      },
+      options
+    );
+
+    const message = this.supportsOpMsg
+      ? new OpMsgRequest(ns.db, cmd, commandOptions)
+      : new OpQueryRequest(ns.db, cmd, commandOptions);
+
+    try {
+      write(this as any as Connection, message, commandOptions, callback);
+    } catch (err) {
+      callback(err);
+    }
+  }
+}
+
+const kDefaultMaxBsonMessageSize = 1024 * 1024 * 16 * 4;
+
+/**
+ * @internal
+ *
+ * This helper reads chucks of data out of a socket and buffers them until it has received a
+ * full wire protocol message.
+ *
+ * By itself, produces an infinite async generator of wire protocol messages and consumers must end
+ * the stream by calling `return` on the generator.
+ *
+ * Note that `for-await` loops call `return` automatically when the loop is exited.
+ */
+export async function* readWireProtocolMessages(
+  connection: ModernConnection
+): AsyncGenerator<Buffer> {
+  const bufferPool = new BufferPool();
+  const maxBsonMessageSize = connection.hello?.maxBsonMessageSize ?? kDefaultMaxBsonMessageSize;
+  for await (const [chunk] of on(connection.socket, 'data')) {
```

Comment:
`on` increases our footprint of needed dependencies from `events`. We went out of our way to import `events` in a nested path in FLE to avoid impacting our work to get the driver operational in other runtimes.

I believe we could use readable's asyncIterator method reducing the contract for what we need out of a socket. We can also follow up with an improvement if it is overly difficult to achieve.